primitives/src/witness.rs

```diff
@@ -257,7 +255,14 @@ fn encode_cursor(bytes: &mut [u8], start_of_indices: usize, index: usize, value:
 #[inline]
 fn decode_cursor(bytes: &[u8], start_of_indices: usize, index: usize) -> Option<usize> {
     let start = start_of_indices + index * 4;
-    bytes.get_array::<4>(start).map(|index_bytes| u32::from_ne_bytes(*index_bytes) as usize)
+    let pos = bytes.get_array::<4>(start).map(|index_bytes| u32::from_ne_bytes(*index_bytes))?;
+    // On a 16-bit machine the u32 _should_ be smaller than `u16::MAX` since one cannot
+    // create a vector bigger than the amount of available memory, check just to be defensive.
+    if mem::size_of::<usize>() <= 2 && pos > u32::from(u16::MAX) {
+        None
+    } else {
+        Some(pos as usize)      // Cast ok, just checked we have at least 32 bits.
```

Comment:
In f94a8a92d2f58ff7e5b46e1c89b4468373266e2b:

Why not just use `usize::try_from`? This logic is tortuous and runs through the definition of `usize`.

Reply:
In f94a8a92d2f58ff7e5b46e1c89b4468373266e2b:

This commit is unchanged in the latest force-push.